llvm/lib/Transforms/IPO/LowerTypeTests.cpp

```diff
@@ -1471,6 +1471,11 @@ void LowerTypeTestsModule::createJumpTable(
   SmallVector<Value *, 16> AsmArgs;
   AsmArgs.reserve(Functions.size() * 2);
 
+  // Check if the first entry has the NoUnwind attribute.
```

Comment:
I don't understand why the first entry is treated specially.

I'd have expected that if _any_ function in the dispatch table might throw an exception through it, then the dispatch table would need unwind info. So I'd expect to see something more like "if _all_ functions have NoUnwind then give the dispatch table NoUnwind as well". Why just the first?

Reply:
You raise a good point and my understanding of this issue specifically crops up when these `cfi.jumptable` functions are replaced as direct calls. AFAIK there won't be direct call replacements for dispatch tables greater than size 1, but I would like to double check to confirm this.

Reply:
I confirmed that with dispatch tables of size > 1 I do not see direct calls to the jump table for throwing functions: 
```llvm
define dso_local void @_Z10call_catchi(i32 noundef %num) local_unnamed_addr #3 personality ptr @__gxx_personality_v0 !type !4 !type !5 !type !6 {
entry:
  store ptr getelementptr inbounds ([2 x [8 x i8]], ptr @_Z8no_throwi.cfi_jt, i64 0, i64 1), ptr @catch_ptr, align 8, !tbaa !14
  invoke void getelementptr inbounds ([2 x [8 x i8]], ptr @_Z8no_throwi.cfi_jt, i64 0, i64 1)(i32 noundef %num) #8
          to label %try.cont unwind label %lpad

lpad:                                             ; preds = %entry
  %0 = landingpad { ptr, i32 }
          catch ptr @_ZTIi
  %1 = extractvalue { ptr, i32 } %0, 1
  %2 = tail call i32 @llvm.eh.typeid.for(ptr nonnull @_ZTIi) #10
  %matches = icmp eq i32 %1, %2
  br i1 %matches, label %catch, label %eh.resume

catch:                                            ; preds = %lpad
  %3 = extractvalue { ptr, i32 } %0, 0
  %4 = tail call ptr @__cxa_begin_catch(ptr %3) #10
  tail call void @__cxa_end_catch() #10
  br label %try.cont

try.cont:                                         ; preds = %catch, %entry
  ret void

eh.resume:                                        ; preds = %lpad
  resume { ptr, i32 } %0
}
```

Reply:
In that case, can you add a comment in the patch explaining all this, please? If this confused me, it will probably confuse other people reading the code, and they shouldn't have to find their way to this PR to get the explanation.

Reply:
Perhaps there's a way to add an assertion / a group of assertions to be paired with the comment? (In case something causes this property to change in the future?)

Reply:
I wonder if its even worth bothering with this special case and just unequivocally removing NoUnwind for the `cfi.jumptable` function. We would lose out on the optimization of removing unwind information only when:

1. A function marked NoUnwind is used as an indirect function
2. It is the only indirect function with that type signature in the link time unit
3. The link time unit is built without `-fsanitize-cfi-cross-dso`
4. The `cfi.jumptable` function gets directly replaced for that indirect call

I don't think the added code complexity is worth this rare optimization.

Reply:
Your idea (to remove the optimization in that case) sounds plausible to me, but I don't really know. Might be good to get some broader input.

Reply:
CC: @efriedma-quic who added this optimization in https://reviews.llvm.org/D46082

Reply:
I haven't looked at this in a while, but I think the issue I was running into was that the EH info was causing issues for a linker script that was expecting no EH info.  Currently, the only way to suppress the generation of the ".cfi_startproc" (which genreates EH info) is "nounwind", I think.

Ideally, we would skip emitting EH info for all jump tables functions; even if the callee unwinds, the jump table isn't on the stack at that point anyway.  Maybe worth adding some way to represent that.

Reply:
re: checking only the first entry, please consider the semantics in general, not just the specific optimizations which exist at the moment.  Even if we don't optimize tables with multiple entries at the moment, someone could try to implement such an optimization in the future.

Reply:
> Ideally, we would skip emitting EH info for all jump tables functions; even if the callee unwinds, the jump table isn't on the stack at that point anyway. Maybe worth adding some way to represent that.

IIUC we'd be able to differentiate NoUnwind from "don't emit EH info" for cfi jump table functions? 

> re: checking only the first entry, please consider the semantics in general, not just the specific optimizations which exist at the moment. Even if we don't optimize tables with multiple entries at the moment, someone could try to implement such an optimization in the future.

I appreciate this, I think its better to be safe in this instance. I will update the patch to instead check if all functions are marked NoUnwind.

Reply:
>IIUC we'd be able to differentiate NoUnwind from "don't emit EH info" for cfi jump table functions?

Right.  Usually, whether unwinding is possible and whether unwind info is required are strongly correlated.  But if you have a function that tail-calls a function that unwinds, it's possible to unwind "through" the function without unwind info.